src/Endpoints/content-helper/class-smart-linking-endpoint.php

```diff
@@ -0,0 +1,267 @@
+				Base_Endpoint::DEFAULT_ACCESS_CAPABILITY
+			)
+		);
+
+		return $can_edit && $has_capability;
+	}
+
+	/**
+	 * Registers the endpoints.
+	 *
+	 * @since 3.16.0
+	 */
+	public function run(): void {
+		$this->register_endpoint(
+			static::ENDPOINT . '/url-to-post-type',
+			'url_to_post_type',
+			array( 'POST' )
+		);
+
+		$this->register_endpoint_with_args(
+			static::ENDPOINT . '/(?P<post_id>\d+)/add',
+			'add_smart_link',
+			array( 'POST' )
+		);
+	}
+
+	/**
+	 * Converts a URL to a post type.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @return WP_REST_Response The response object.
+	 */
+	public function url_to_post_type( WP_REST_Request $request ): WP_REST_Response {
+		$url = $request->get_param( 'url' );
+
+		if ( ! is_string( $url ) ) {
+			return new WP_REST_Response(
+				array(
+					'error' => array(
+						'name'    => 'invalid_request',
+						'message' => 'Invalid request body.',
+					),
+				),
+				400
+			);
+		}
+
+		$post_id = 0;
+		$cache   = wp_cache_get( $url, 'wp_parsely_smart_link_url_to_postid' );
+
+		if ( is_integer( $cache ) ) {
+			$post_id = $cache;
+		} elseif ( function_exists( 'wpcom_vip_url_to_postid' ) ) {
+			$post_id = wpcom_vip_url_to_postid( $url );
+		} else {
+			// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.url_to_postid_url_to_postid
+			$post_id = url_to_postid( $url );
+			wp_cache_set( $url, $post_id, 'wp_parsely_smart_link_url_to_postid' );
+		}
+
+		$response = array(
+			'data' => array(
+				'post_id'   => false,
+				'post_type' => false,
+			),
+		);
+
+		if ( 0 !== $post_id ) {
+			$response['data']['post_id']   = $post_id;
+			$response['data']['post_type'] = get_post_type( $post_id );
+		}
+
+		return new WP_REST_Response( $response, 200 );
+	}
+
+	/**
+	 * Validates the request parameters.
+	 *
+	 * @throws InvalidArgumentException If the request is invalid.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @return bool True if the request is valid, false otherwise.
+	 */
+	private function validate_request( WP_REST_Request $request ): bool {
+		$post_id = $request->get_param( 'post_id' );
+
+		if ( ! is_numeric( $post_id ) ) {
+			throw new InvalidArgumentException( 'Invalid post ID' );
+		}
+
+		// Check if post exists.
+		$post = get_post( intval( $post_id ) );
+		if ( null === $post ) {
+			throw new InvalidArgumentException( 'Invalid post ID' );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Adds a smart link to a post.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 * @return WP_REST_Response The response object.
+	 */
+	public function add_smart_link( WP_REST_Request $request ): WP_REST_Response {
+		try {
+			$this->validate_request( $request );
+
+			$post_id = $request->get_param( 'post_id' );
+			if ( is_numeric( $post_id ) ) {
+				$post_id = intval( $post_id );
+			} else {
+				// Validate makes sure it's numeric, so this should never happen.
+				$post_id = 0;
+			}
```

Comment:
Should we leave this code as an extra measure, or maybe we want to handle this in any other way? Fine either way.

Reply:
This was basically to make PHPStan happy. Will be improved on the next PR 🙂